### PR TITLE
Add experimental speech translation and voice catalog APIs

### PR DESCRIPTION
**Pull request by ttupper92618**
## Summary

- add experimental `POST /v1/audio/translations` over the existing bounded STT path
- map English-target translation intent to model-family runner arguments without advertising an unvalidated model
- add the Skulk `GET /v1/audio/voices` extension and static model-card voice catalogs
- add the `speech_translation` experiment setting and dashboard control
- document the APIs, schema, settings, and architecture

## Validation

- `uv run basedpyright`
- `uv run ruff check`
- `nix fmt`
- `uv run pytest` (2209 passed, 1 skipped)
- `npm run build --prefix dashboard-react`
- `uv run python scripts/build_docs.py`

### CODE REVIEW

File: src/skulk/api/main.py [hunk 1]
```diff
@@ -9976,6 +10038,94 @@ async def audio_transcriptions(
...
+            File(description="Audio file to translate to English."),
+        ],
+        model: Annotated[
+            str,
+            Form(description="Mounted translation-capable speech model id."),
+        ],
+        language: Annotated[
+            str | None,
+            Form(description="Required model-specific source language code."),
+        ] = None,
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:10072]: Corrected the FastAPI form-field description to say the source language is optional, matching the type and API guide. Focused translation API tests, basedpyright, and Ruff pass.

File: src/skulk/api/main.py [hunk 2]
```diff
@@ -3374,6 +3414,28 @@ async def _validate_audio_transcription_model(self, model_id: ModelId) -> ModelI
...
+        if not experimental_mode_enabled() or not self._speech_translation_experiment_enabled():
+            raise HTTPException(
+                status_code=400,
+                detail=(
+                    "Speech translation is experimental and requires "
+                    f"{EXPERIMENTAL_MODE_ENV_VAR}=1 and "
+                    "experiments.speech_translation=true"
+                ),
+            )
+        model_card = await self._get_running_model_card(model_id)
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:3442]: Fixed by requiring a mounted instance before translation validation succeeds. Unmounted cards now trigger the normal notification path and return 404; focused API tests cover rejection before dispatch.

File: src/skulk/api/main.py [hunk 3]
```diff
@@ -9976,6 +10038,94 @@ async def audio_transcriptions(
...
+        model: Annotated[
+            str,
+            Query(description="Mounted text-to-speech model id."),
+        ],
+    ) -> AudioVoiceList:
+        """Return built-in voices declared by one mounted TTS model."""
+
+        if not model.strip():
+            raise HTTPException(status_code=400, detail="`model` must not be empty")
+        model_card = await self._get_running_model_card(ModelId(model))
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py]: Fixed by requiring mounted capacity before returning a voice catalog. Added a focused test proving an unmounted catalog card returns 404.

File: src/skulk/api/main.py [hunk 4]
```diff
@@ -3374,6 +3414,28 @@ async def _validate_audio_transcription_model(self, model_id: ModelId) -> ModelI
...
+            )
+        model_card = await self._get_running_model_card(model_id)
+        resolved = model_card.model_id
+        profile = resolve_model_capability_profile(resolved, model_card=model_card)
+        if not profile.supports_speech_translation:
+            raise HTTPException(
+                status_code=400,
+                detail=f"Model {resolved} does not support speech translation",
+            )
+        return resolved
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:3455]: Fixed the duplicated translation availability finding with the mounted-instance check and focused 404 coverage.

File: src/skulk/api/main.py [hunk 5]
```diff
@@ -9976,6 +10038,94 @@ async def audio_transcriptions(
...
+            raise HTTPException(
+                status_code=400,
+                detail=f"Model {resolved} does not support voice listing",
+            )
+        return AudioVoiceList(
+            data=tuple(
+                AudioVoice(id=voice, name=voice, model=str(resolved))
+                for voice in model_card.audio.voices
+            )
+        )
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py]: Fixed the duplicated voice availability finding with the mounted-instance check and focused 404 coverage.

File: src/skulk/api/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:10072]: The description was corrected in the preceding commit to say the source language is optional, matching the type and API guide.

File: src/skulk/api/main.py [hunk 6]
```diff
@@ -9976,6 +10038,94 @@ async def audio_transcriptions(
...
+            raise HTTPException(status_code=400, detail="`model` must not be empty")
+        if response_format not in _AUDIO_TRANSCRIPTION_FORMATS:
+            raise HTTPException(
+                status_code=400,
+                detail=f"Unsupported translation response_format: {response_format}",
+            )
+        _validate_audio_upload_metadata(file)
+        audio_bytes = await _read_audio_upload(file)
+        model_id = await self._validate_audio_translation_model(ModelId(model))
+        transcript_chunks = await self._execute_audio_transcription(
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:10100]: Moved translation model/experiment validation before upload metadata validation and body consumption. Added a test asserting rejected requests leave the upload unread.

File: src/skulk/api/main.py [hunk 7]
```diff
@@ -9976,6 +10047,103 @@ async def audio_transcriptions(
...
+        model: Annotated[
+            str,
+            Query(description="Mounted text-to-speech model id."),
+        ],
+    ) -> AudioVoiceList:
+        """Return built-in voices declared by one mounted TTS model."""
+
+        if not model.strip():
+            raise HTTPException(status_code=400, detail="`model` must not be empty")
+        model_card = await self._get_running_model_card(ModelId(model))
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py]: Fixed by checking State for the requested mounted model before calling `_get_running_model_card()`. Unknown/unmounted voice requests now return 404 without Hub lookup or custom-card writes; the focused test makes card lookup fail if invoked.

File: src/skulk/worker/runner/speech/runner.py [hunk 8]
```diff
@@ -138,6 +139,13 @@ def _stt_generate_kwargs(
         }
     return {
         "language": params.language,
+        "task": "translate" if params.translate_to_english else None,
+        "source_lang": params.language,
+        "target_lang": "en" if params.translate_to_english else params.language,
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/runner/speech/runner.py]: Fixed by omitting the ambiguous generic `language` alias whenever `translate_to_english` is active, while retaining explicit `source_lang=fr` and `target_lang=en`. Whisper-style translation keeps its separate `language` plus `task=translate` branch. Added focused coverage for variadic Canary-style generation APIs; 20 speech runner tests, basedpyright, Ruff, and formatting pass.

File: src/skulk/worker/runner/speech/runner.py [hunk 9]
```diff
@@ -104,6 +105,40 @@ def _filter_kwargs(fn: Callable[..., Any], kwargs: dict[str, Any]) -> dict[str,
...
+def _install_canary_compatibility() -> None:
+    """Install runtime compatibility required by upstream bfloat16 Canary."""
+
+    from mlx_audio.stt.models.canary.decoder import (
+        MultiHeadCrossAttention,
+        MultiHeadSelfAttention,
+    )
+
+    _install_attention_mask_dtype_compat(MultiHeadSelfAttention)
+    _install_attention_mask_dtype_compat(MultiHeadCrossAttention)
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/runner/speech/runner.py]: Fixed in the current head. Canary cross-attention now has a dedicated compatibility implementation preserving `(x, encoder_output, encoder_mask, cache)` and constructs its internal mask in the projected query dtype; self-attention retains the generic adapter. A focused bfloat16 cross-attention test passes, and repeated live French-to-English translations completed successfully.

File: src/skulk/shared/models/model_cards.py [hunk 10]
```diff
@@ -383,6 +385,26 @@ def _coerce_sample_rates(cls, value: object) -> tuple[int, ...]:
...
+        if not isinstance(value, Iterable) or isinstance(value, (str, bytes)):
+            raise ValueError("voices must be a list of non-empty identifiers")
+        voices = tuple(
+            str(item).strip() for item in cast("Iterable[object]", value)
+        )
+        if any(not voice for voice in voices):
+            raise ValueError("voices must contain only non-empty identifiers")
+        if len(set(voices)) != len(voices):
+            raise ValueError("voices must not contain duplicates")
+        return voices
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/shared/models/model_cards.py:402]: Severity 2 (low): model-card validation already rejects blank and duplicate voice identifiers, and this malformed Python-input edge requires a programmatic/admin card author rather than a normal API workload. Per the repository severity policy, this is deferred rather than expanding the current API slice.

File: website/docs/architecture.md [hunk 11]
```diff
@@ -341,8 +341,12 @@ claim VAD, conversation, or speech-to-speech behavior. Dashboard chat selects
...
 24 kHz PCM16 contract, while non-realtime models retain batch MediaRecorder
-transcription. Voice/reference-audio management and speech translation remain
-later phases.
+transcription. Mounted TTS cards with static `audio.voices` metadata expose it
+through the Skulk `GET /v1/audio/voices` extension. Translation-capable STT
+cards can reuse the bounded batch path through experimental
+`POST /v1/audio/translations`; the speech runner maps the English-target intent
+to model-family generation arguments. Reference-audio management remains a
+later phase.
 See [Speech Fabric and Realtime Design](speech-fabric-realtime).
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply website/docs/architecture.md:350]: Severity 1 (nit): the link target is correct and the existing label remains descriptive, so this wording-only rename is deferred under the review severity policy.

[the remark this answers is not in the document]
Comment (ttupper92618) [reply website/docs/architecture.md:350]: Severity 1 duplicate of the wording-only link-label suggestion. The target resolves correctly; deferred under the review severity policy.

File: src/skulk/api/main.py [hunk 12]
```diff
@@ -9976,6 +10047,103 @@ async def audio_transcriptions(
...
+                detail=f"No instance found for model {requested_model}",
+            )
+        model_card = await self._get_running_model_card(requested_model)
+        resolved = model_card.model_id
+        profile = resolve_model_capability_profile(resolved, model_card=model_card)
+        if not profile.supports_speech_synthesis:
+            raise HTTPException(
+                status_code=400,
+                detail=f"Model {resolved} is not a text-to-speech model",
+            )
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:10142]: Added focused coverage proving a mounted STT model returns 400 from the voice catalog rather than being treated as TTS.

File: src/skulk/api/main.py [hunk 13]
```diff
@@ -9976,6 +10047,103 @@ async def audio_transcriptions(
...
+        if (
+            model_card.audio is None
+            or model_card.audio.supports_voice_listing is not True
+        ):
+            raise HTTPException(
+                status_code=400,
+                detail=f"Model {resolved} does not support voice listing",
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:10149]: Added focused coverage proving a mounted TTS card without explicit voice-listing support returns the documented 400.

File: src/skulk/api/main.py [hunk 14]
```diff
@@ -3374,6 +3414,37 @@ async def _validate_audio_transcription_model(self, model_id: ModelId) -> ModelI
...
+
+        if not experimental_mode_enabled() or not self._speech_translation_experiment_enabled():
+            raise HTTPException(
+                status_code=400,
+                detail=(
+                    "Speech translation is experimental and requires "
+                    f"{EXPERIMENTAL_MODE_ENV_VAR}=1 and "
+                    "experiments.speech_translation=true"
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:3432]: Added focused coverage proving the translation experiment gate returns 400 before consuming the upload body.

File: src/skulk/api/main.py [hunk 15]
```diff
@@ -3374,6 +3414,37 @@ async def _validate_audio_transcription_model(self, model_id: ModelId) -> ModelI
...
+                detail=f"No instance found for model {model_id}",
+            )
+        model_card = await self._get_running_model_card(model_id)
+        resolved = model_card.model_id
+        profile = resolve_model_capability_profile(resolved, model_card=model_card)
+        if not profile.supports_speech_translation:
+            raise HTTPException(
+                status_code=400,
+                detail=f"Model {resolved} does not support speech translation",
+            )
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:3449]: Added focused coverage proving mounted transcription-only capacity without translation capability returns 400.

File: src/skulk/worker/runner/speech/tests/test_speech_runner.py [hunk 16]
```diff
@@ -262,6 +265,128 @@ def _generate(
...
+
+
+def test_canary_cross_attention_accepts_bfloat16_with_encoder_mask() -> None:
+    """Canary cross-attention must keep its internally built mask in query dtype."""
+
+    import mlx.core as mx
+    from mlx.utils import tree_map
+    from mlx_audio.stt.models.canary.decoder import (  # pyright: ignore[reportMissingTypeStubs]
+        MultiHeadCrossAttention,
+    )
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/runner/speech/tests/test_speech_runner.py:320]: Fixed by guarding both optional imports with `pytest.importorskip`. Non-MLX environments now skip only this backend-specific regression test; the focused speech suite, basedpyright, Ruff, and formatting pass.

File: website/docs/architecture.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (ttupper92618) [reply website/docs/architecture.md:350]: Severity 1 wording-only duplicate. The link target and descriptive label are both valid; deferred under the repository review severity policy.

File: src/skulk/worker/runner/speech/runner.py [hunk 17]
```diff
@@ -137,7 +222,16 @@ def _stt_generate_kwargs(
             "verbose": True,
         }
     return {
-        "language": params.language,
+        # Canary treats ``language`` as both source and target, so emitting the
+        # generic alias here would overwrite the explicit English target.
+        "language": None if params.translate_to_english else params.language,
+        "task": "translate" if params.translate_to_english else None,
+        "source_lang": params.language,
+        "target_lang": "en" if params.translate_to_english else params.language,
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/worker/runner/speech/runner.py:230]: Fixed by passing the source language into translation model validation and requiring it for Canary before upload metadata/body consumption. Added focused 400 coverage and documented the family-specific requirement; 20 translation API tests, basedpyright, Ruff, and formatting pass.

File: src/skulk/api/main.py [hunk 18]
```diff
@@ -9976,6 +10047,103 @@ async def audio_transcriptions(
...
+        model_id = await self._validate_audio_translation_model(ModelId(model))
+        _validate_audio_upload_metadata(file)
+        audio_bytes = await _read_audio_upload(file)
+        transcript_chunks = await self._execute_audio_transcription(
+            AudioTranscriptionTaskParams(
+                model=model_id,
+                filename=file.filename,
+                content_type=_normalize_upload_content_type(file.content_type),
+                audio_sha256=hashlib.sha256(audio_bytes).hexdigest(),
```
[the remark this answers is not in the document]
Comment (ttupper92618) [reply src/skulk/api/main.py:10105]: Severity 2 (low): this is bounded duplicate hashing on an experimental endpoint and has no correctness impact. The authoritative checksum is still recomputed in the shared execution path. Deferred under the repository review severity policy.
